### PR TITLE
Add partitioned nvFuser executor with ATen fallbacks

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR introduces a new nvFuser executor for FX graphs containing different kinds of nodes, not just `torch.ops.prims` supported by nvFuser. The FX graph is partitioned based on whether nodes are supported or not by nvFuser and supported nodes are fused into subgraphs, that's all using Sherlock's work on the partitioner.

This new partitions-based executor with fallbacks to ATen is used by default with `executor="nvfuser"`. And the previous executor can be used with `executor="strictly_nvfuser"`, naming suggestions are welcome!

### CODE REVIEW

File: torch/_prims/nvfuser_executor.py [hunk 1]
```diff
@@ -119,3 +121,57 @@ def nvfuser_execute(gm: GraphModule, *args):
...
+    any_unsupported = any(
+        not supported_ops.is_node_supported(None, node) for node in call_function_nodes
+    )
+    if any_unsupported:
+        # CapabilityBasedPartitioner modifies the graph in-place so we need to make a copy of the graph
+        gm = deepcopy(gm)
+        partitioner = CapabilityBasedPartitioner(
+            gm, supported_ops, allows_single_node_partition=True
+        )
+        partitions = partitioner.propose_partitions()
```
Comment (SherlockNoMad) [inline torch/_prims/nvfuser_executor.py:163]: FYI, partitioner.partition_and_fuse() does these two line

**Comment by SherlockNoMad** [comment]
Hi @IvanYashchuk, this PR assumes that gm module passed to executor has been lower to prims... 

but the main question is when should we perform aten->aten->prims decomp? see discussion here https://github.com/pytorch/pytorch/issues/80541 and https://github.com/pytorch/pytorch/issues/80966

We want to determine the decomposition earlier in the stack, so that if a decomposition cannot be fused latter, it's still able to fallback to the original "fused-aten" op. 

In light of this, executor will always get a subgraph that is fully fusible. 

So I would like to get your perspective on what's the benefit of deferring the partitioner to executor?

File: torch/_prims/nvfuser_executor.py [hunk 2]
```diff
@@ -119,3 +121,57 @@ def nvfuser_execute(gm: GraphModule, *args):
...
+    else:
+        return gm, any_unsupported
+
+
+def nvfuser_execute_partitioned(gm: GraphModule, *args):
+    # When possible it's better to use nvfuser_execute directly
+    # because it avoids PartitionedInterpreter's overhead
+    gm, is_partitioned = maybe_partition_graph(gm)
+    if is_partitioned:
+        return PartitionedInterpreter(gm).run(*args)
```
Comment (IvanYashchuk) [inline torch/_prims/nvfuser_executor.py:175]: We should check that there's at least one node runnable by `nvfuser_execute`.

File: torch/_prims/nvfuser_executor.py (hunk 1 above)
Comment (IvanYashchuk) [reply torch/_prims/nvfuser_executor.py:163]: Thanks, I know about this function. I wanted to check the length of the proposed partitions and then forgot. There should be an explicit error before trying to construct empty nvFuser's Fusion object.

File: torch/_prims/nvfuser_executor.py [hunk 3]
```diff
@@ -119,3 +121,61 @@ def nvfuser_execute(gm: GraphModule, *args):
...
+    )
+    if any_unsupported:
+        # CapabilityBasedPartitioner modifies the graph in-place so we need to make a copy of the graph
+        gm = deepcopy(gm)
+        partitioner = CapabilityBasedPartitioner(
+            gm, supported_ops, allows_single_node_partition=True
+        )
+        partitions = partitioner.propose_partitions()
+        if len(partitions) == 0:
+            raise RuntimeError(
```
Comment (jjsjann123) [inline torch/_prims/nvfuser_executor.py:163]: nitpick, error out feels like a confusing UX. Maybe change this to a warning instead?

File: test/test_prims.py [hunk 4]
```diff
@@ -149,6 +149,58 @@ def func(a):
...
+        a = torch.randn(3, 4, device=device)
+
+        def func(a):
+            return torch.digamma(a)  # not supported by nvfuser
+
+        with TorchRefsMode.push():
+            gm = make_fx(func)(a)
+
+        with self.assertRaisesRegex(RuntimeError, "is not supported by nvFuser"):
+            execute(gm, a, executor="nvfuser")
```
Comment (jjsjann123) [inline test/test_prims.py:202]: ditto, feels a little bit strange when the executor that supports aten fallback errors out.

**Review by jjsjann123 — APPROVED** [review]
LGTM, minor nitpick on assertion.

File: test/test_prims.py (hunk 4 above)
Comment (IvanYashchuk) [reply test/test_prims.py:202]: Agreed, I will change it to emit a warning instead.

Comment (IvanYashchuk) [reply test/test_prims.py:235]: Done.

**Review by SherlockNoMad — APPROVED** [review]

**Pull request closed without merging**